newrelic_python_agent/agent.py

```diff
@@ -104,49 +108,113 @@ def license_key(self):
+        :param dict instance: The instance config block
+        :rtype: str
+        """
+        # start with the base name of the plugin + name field
+        name = "%s:%s" % (plugin_name, instance.get('name', 'unnamed'))
+        i = 0
+        instance_name = "%s:%i" % (name, i)
+        while instance_name in self.thread_names:
+            i = i + 1
+            instance_name = "%s:%i" % (name, i)
+
+        return instance_name
+
+    def start_plugin(self, plugin_name, plugin, config):
         """Kick off a background thread to run the processing task.
 
-        :param newrelic_python_agent.plugins.base.Plugin plugin: The plugin
-        :param dict config: The config for the plugin
+        :param plugin: The plugin name as defined in the application config
+        :param config: The set of instance configs for the plugin
+        :type plugin: newrelic_python_agent.plugins.base.Plugin
+        :type config: dict or list(dict)
 
         """
 
         if not isinstance(config, (list, tuple)):
             config = [config]
 
+        LOGGER.debug("Plugin config: %s", config)
+
+        # the instance names must be unique so we can store the results for each.
+        # use the 'name' field, if specified.  If there are duplicate names, we
+        # simply append a number so it is unique.  As long as the the order in the
+        # config remains the same, then the instance number will remain the same.
         for instance in config:
-            thread = threading.Thread(target=self.thread_process,
-                                      kwargs={'config': instance,
-                                              'name': plugin_name,
-                                              'plugin': plugin,
-                                              'poll_interval':
-                                                  int(self._wake_interval)})
-            thread.start()
+            instance_name = self.get_instance_name(plugin_name, instance)
+
+            if issubclass(plugin, base.ConfigPlugin):
+                thread = threading.Thread(target=self.thread_config_process,
+                                          kwargs={'config': instance,
+                                                  'name': instance_name,
+                                                  'plugin': plugin})
+            else:
+                thread = threading.Thread(target=self.thread_metric_process,
+                                          kwargs={'config': instance,
+                                                  'name': instance_name,
+                                                  'plugin': plugin,
+                                                  'poll_interval':
+                                                      int(self._wake_interval)})
+            LOGGER.info("Starting plugin instance %s as thread %s", instance_name, thread.getName())
+            self.thread_names[instance_name] = thread.getName()
+            thread.run()
```

Comment:
argh.  looks like i messed up the rebase of this PR and accidentally reverted the use_threading PR.  I'll fix this.